utils/iamclient/iamclient.go

```diff
@@ -0,0 +1,1139 @@
+
+		return aoserrors.Wrap(err)
+	}
+
+	client.nodeInfoSubs.grpcStream = &listener
+
+	client.nodeInfoSubs.stopWG.Add(1)
+	go client.processNodeInfoChange(client.nodeInfoSubs)
+
+	return nil
+}
+
+func (client *Client) subscribeUnitSubjectsChange() error {
+	listener, err := client.identService.SubscribeSubjectsChanged(context.Background(), &emptypb.Empty{})
+	if err != nil {
+		log.WithField("error", err).Errorf("Can't subscribe on unit subjects change event: %v", err)
+
+		return aoserrors.Wrap(err)
+	}
+
+	client.subjectsSubs.grpcStream = &listener
+
+	client.subjectsSubs.stopWG.Add(1)
+	go client.processUnitSubjectsChange(client.subjectsSubs)
+
+	return nil
+}
+
+func (client *Client) subscribeCertChange(certType string) (
+	listener pb.IAMPublicService_SubscribeCertChangedClient, err error,
+) {
+	listener, err = client.publicService.SubscribeCertChanged(context.Background(),
+		&pb.SubscribeCertChangedRequest{Type: certType})
+	if err != nil {
+		log.WithField("error", err).Error("Can't subscribe on CertChange event")
+
+		return nil, aoserrors.Wrap(err)
+	}
+
+	return listener, aoserrors.Wrap(err)
+}
+
+func (client *Client) processNodeInfoChange(sub *nodeInfoChangeSub) {
+	defer sub.stopWG.Done()
+
+	for {
+		nodeInfo, err := (*sub.grpcStream).Recv()
+		if err != nil {
+			log.WithField("err", err).Error("Process NodeInfo change failed")
+			client.onConnectionLost()
+
+			return
+		}
+
+		client.Lock()
+		for _, listener := range sub.listeners {
+			listener <- pbconvert.NodeInfoFromPB(nodeInfo)
+		}
+		client.Unlock()
+	}
+}
+
+func (client *Client) processUnitSubjectsChange(sub *subjectsChangeSub) {
+	defer sub.stopWG.Done()
+
+	for {
+		subjects, err := (*sub.grpcStream).Recv()
+		if err != nil {
+			log.WithField("err", err).Error("Process UnitSubjects change failed")
+			client.onConnectionLost()
+
+			return
+		}
+
+		client.Lock()
+		for _, listener := range sub.listeners {
+			listener <- subjects.GetSubjects()
+		}
+		client.Unlock()
+	}
+}
+
+func (client *Client) processCertInfoChange(sub *certChangeSub) {
+	log.Debug("Start process CertInfo change")
+
+	defer sub.stopWG.Done()
+
+	for {
+		cert, err := (*sub.grpcStream).Recv()
+		if err != nil {
+			log.WithField("err", err).Error("Process CertInfo change failed")
+
+			client.onConnectionLost()
+
+			return
+		}
+
+		client.Lock()
+		for _, listener := range sub.listeners {
+			listener <- cert
+		}
+		client.Unlock()
+	}
+}
+
+func (client *Client) reconnect() {
+	defer client.reconnectWG.Done()
+	defer log.Debug("Reconnecting to IAM server...")
+
+	// Give some time for grpc channels to disconnect before first reconnect attempt
+	timer := time.NewTimer(iamFirstConnectAttemptDelay)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-client.closeChannel:
+			// don't recover reconnecting state if client is closing
+			client.isReconnecting.Store(true)
+
+			return
+		case <-timer.C:
+			client.Lock()
```

Comment:
Why lock is required? Deadlock can occur if reconnect happens during some rpc call. Right?

Reply:
lock required, because reconnection changes *service instances of the client, utilized by grpc calls. Also required to sync with Close method, that invokes closeGRPCConnection() from another thread